zcash_client_sqlite/src/wallet.rs

```diff
@@ -891,48 +891,107 @@ pub(crate) fn update_expired_notes<P>(
 ///
 /// This is a crate-internal convenience method.
-pub(crate) fn insert_sent_output<'a, P: consensus::Parameters>(
-    stmts: &mut DataConnStmtCache<'a, P>,
+pub(crate) fn insert_sent_output<P: consensus::Parameters>(
+    conn: &Connection,
+    params: &P,
     tx_ref: i64,
     from_account: AccountId,
     output: &SentTransactionOutput,
 ) -> Result<(), SqliteClientError> {
-    stmts.stmt_insert_sent_output(
-        tx_ref,
-        output.output_index(),
-        from_account,
-        output.recipient(),
-        output.value(),
-        output.memo(),
-    )
+    let mut stmt_insert_sent_output = conn.prepare_cached(
+        "INSERT INTO sent_notes (
+            tx, output_pool, output_index, from_account,
+            to_address, to_account, value, memo)
+        VALUES (
+            :tx, :output_pool, :output_index, :from_account,
+            :to_address, :to_account, :value, :memo)",
+    )?;
+
+    let (to_address, to_account, pool_type) = recipient_params(params, output.recipient());
+    let sql_args = named_params![
+        ":tx": &tx_ref,
+        ":output_pool": &pool_code(pool_type),
+        ":output_index": &i64::try_from(output.output_index()).unwrap(),
+        ":from_account": &u32::from(from_account),
+        ":to_address": &to_address,
+        ":to_account": &to_account,
+        ":value": &i64::from(output.value()),
+        ":memo": output.memo().filter(|m| *m != &MemoBytes::empty()).map(|m| m.as_slice()),
+    ];
+
+    stmt_insert_sent_output.execute(sql_args)?;
+
+    Ok(())
 }
 
 /// Records information about a transaction output that your wallet created.
 ///
 /// This is a crate-internal convenience method.
 #[allow(clippy::too_many_arguments)]
 pub(crate) fn put_sent_output<'a, P: consensus::Parameters>(
-    stmts: &mut DataConnStmtCache<'a, P>,
+    conn: &Connection,
+    params: &P,
     from_account: AccountId,
     tx_ref: i64,
     output_index: usize,
     recipient: &Recipient,
     value: Amount,
     memo: Option<&MemoBytes>,
 ) -> Result<(), SqliteClientError> {
-    if !stmts.stmt_update_sent_output(from_account, recipient, value, memo, tx_ref, output_index)? {
-        stmts.stmt_insert_sent_output(
-            tx_ref,
-            output_index,
-            from_account,
-            recipient,
-            value,
-            memo,
-        )?;
-    }
+    let mut stmt_upsert_sent_output = conn.prepare_cached(
+        "INSERT INTO sent_notes (
+            tx, output_pool, output_index, from_account,
+            to_address, to_account, value, memo)
+        VALUES (
+            :tx, :output_pool, :output_index, :from_account,
+            :to_address, :to_account, :value, :memo)
+        ON CONFLICT (tx, output_pool, output_index) DO UPDATE
+        SET from_account = :from_account,
+            to_address = :to_address,
+            to_account = :to_account,
+            value = :value,
+            memo = IFNULL(:memo, memo)",
```

Comment:
Check whether this has the same issue as https://github.com/zcash/librustzcash/pull/857/commits/c7f759929cd3157039842cd0a4a04c8dc648dc21#r1224660086